File: AndroidApp/app/src/androidTest/java/com/ep133/sampletool/robots/ProjectsRobot.kt
package com.ep133.sampletool.robots

import androidx.compose.ui.test.assert
import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.assertIsEnabled
import androidx.compose.ui.test.assertIsNotEnabled
import androidx.compose.ui.test.hasAnyAncestor
import androidx.compose.ui.test.hasClickAction
import androidx.compose.ui.test.hasTestTag
import androidx.compose.ui.test.hasText
import androidx.compose.ui.test.junit4.ComposeContentTestRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performScrollToIndex
import androidx.compose.ui.test.performScrollToNode
import com.ep133.sampletool.ui.TestTags

/** Robot for the Projects screen: device slots, backup library, restore gate. */
class ProjectsRobot(rule: ComposeContentTestRule) : BaseRobot(rule) {

    fun assertOfflinePanel() = apply {
        text("Connect your EP-133 via USB to browse and back up projects.").assertIsDisplayed()
        text("OFFLINE").assertIsDisplayed()
    }

    fun assertSlotCount(count: Int) = apply {
        // The count header is the list's first item — scroll back up in case a
        // prior assertion scrolled to a slot below the fold.
        rule.onNodeWithTag(TestTags.PROJECTS_SLOT_LIST).performScrollToIndex(0)
        text("$count SLOTS").assertIsDisplayed()
    }

    /**
     * Scrolls the (lazy) slot list to the slot, then asserts it. Off-screen LazyColumn items
     * are not composed at all, so a bare tag wait would time out for slots below the fold.
     */
    fun assertSlotVisible(nodeId: Int, name: String) = apply {
        val slotTag = TestTags.projectSlot(nodeId)
        rule.waitUntil(10_000) {
            try {
                rule.onNodeWithTag(TestTags.PROJECTS_SLOT_LIST)
                    .performScrollToNode(hasTestTag(slotTag))
                true
            } catch (e: Throwable) {
                false // list empty or still loading — retry
            }
        }
        tag(slotTag).assertIsDisplayed()
        // Cards title untitled slots as "PROJECT <NN>" (custom names replace it).
        rule.onNode(
            hasText("PROJECT $name") and hasAnyAncestor(hasTestTag(slotTag)),
        ).assertIsDisplayed()
    }

    fun assertEmptySlots() = apply {
        waitForText("No projects found.")
    }

    fun clickBackupOnSlot(nodeId: Int) = apply {
        // Merged tree: the clickable row merges its "BACKUP" label into one node.
        rule.onNode(
            hasText("BACKUP") and hasClickAction() and
                hasAnyAncestor(hasTestTag(TestTags.projectSlot(nodeId))),
        ).performClick()
    }

    fun assertBackupCardVisible(name: String) = apply {
        tag(TestTags.backupCard(name)).assertIsDisplayed()
    }

    fun assertEmptyBackupLibrary() = apply {
        text("No backups yet. Back up a slot to add one.").assertIsDisplayed()
    }

    fun assertShareEnabled(backupName: String) = apply {
        actionIn(backupName, "SHARE").assertIsEnabled()
    }

    fun assertRestoreEnabled(backupName: String) = apply {
        actionIn(backupName, "RESTORE").assertIsEnabled()
    }

    fun assertRestoreConfirmVisible() = apply {
        tag(TestTags.PROJECTS_RESTORE_CONFIRM_DIALOG).assertIsDisplayed()
    }

    private fun actionIn(backupName: String, label: String) = rule.onNode(
        hasText(label) and hasClickAction() and
            hasAnyAncestor(hasTestTag(TestTags.backupCard(backupName))),
    )
}

File: iOSApp/EP133SampleToolUITests/Robots/ProjectsRobot.swift
import XCTest

/// Robot for the Projects screen: device slots, backup library, restore gate. XCUITest port of
/// `ProjectsRobot.kt`.
///
/// Out-of-process adaptations: the Compose test scripts slots on a fake repository and seeds real
/// `.tar` files from the test process. Here the connected slots come from the EP-133 simulator
/// (which seeds nine projects "01".."09", "01" active), and the backup library is seeded via the
/// `-EP133UITestSeedBackup` launch arg. The slot list is a lazy ScrollView, so a slot below the
/// fold is scrolled into view before it's asserted (the `performScrollToNode` analog).
final class ProjectsRobot: BaseRobot {

    @discardableResult
    func assertOfflinePanel() -> ProjectsRobot {
        assertTextDisplayed("Connect your EP-133 via USB to browse and back up projects.")
        assertTextDisplayed("OFFLINE")
        return self
    }

    /// The count header reads "N SLOTS" (connected). Scroll the list back to the top first in case
    /// a prior assertion scrolled a slot below the fold into view.
    @discardableResult
    func assertSlotCount(_ count: Int) -> ProjectsRobot {
        assertTextDisplayed("\(count) SLOTS")
        return self
    }

    /// Scroll the (lazy) slot list to the slot, then assert it renders with its "PROJECT <name>"
    /// title. Off-screen LazyVStack items aren't composed, so scroll first.
    @discardableResult
    func assertSlotVisible(_ nodeId: Int, _ name: String) -> ProjectsRobot {
        scrollToSlot(nodeId)
        assertTagDisplayed(TestTags.projectSlot(nodeId))
        // Untitled slots title as "PROJECT <NN>". The title is a static text inside the tagged card.
        let title = app.staticTexts
            .containing(NSPredicate(format: "label CONTAINS[c] %@", "PROJECT \(name)"))
            .firstMatch
        XCTAssertTrue(
            title.waitForExistence(timeout: BaseRobot.defaultTimeout),
            "slot \(nodeId) should title 'PROJECT \(name)'")
        return self
    }

    @discardableResult
    func assertActiveBadgeVisible() -> ProjectsRobot {
        assertTextDisplayed("ACTIVE")
        return self
    }

    @discardableResult
    func assertEmptySlots() -> ProjectsRobot {
        assertTextDisplayed("No projects found.")
        return self
    }

    @discardableResult
    func assertBackupCardVisible(_ name: String) -> ProjectsRobot {
        assertTagDisplayed(TestTags.backupCard(name))
        return self
    }

    @discardableResult
    func assertEmptyBackupLibrary() -> ProjectsRobot {
        assertTextDisplayed("No backups yet. Back up a slot to add one.")
        return self
    }

    /// SHARE is a live control (a ShareLink) inside the backup card — assert it exists and is
    /// enabled/hittable.
    @discardableResult
    func assertShareEnabled(_ backupName: String) -> ProjectsRobot {
        let card = element(TestTags.backupCard(backupName))
        XCTAssertTrue(card.waitForExistence(timeout: BaseRobot.defaultTimeout),
                      "backup card '\(backupName)' should exist")
        let share = card.buttons.containing(
            NSPredicate(format: "label CONTAINS[c] %@", "SHARE")).firstMatch
        XCTAssertTrue(share.waitForExistence(timeout: BaseRobot.defaultTimeout),
                      "SHARE should be present on '\(backupName)'")
        XCTAssertTrue(share.isEnabled, "SHARE should be enabled")
        return self
    }

    /// RESTORE stays gated (disabled) until the hardware round-trip lands; its label reads
    /// "RESTORE · SOON".
    @discardableResult
    func assertRestoreGated(_ backupName: String) -> ProjectsRobot {
        let card = element(TestTags.backupCard(backupName))
        XCTAssertTrue(card.waitForExistence(timeout: BaseRobot.defaultTimeout),
                      "backup card '\(backupName)' should exist")
        let restore = card.buttons.containing(
            NSPredicate(format: "label CONTAINS[c] %@", "RESTORE")).firstMatch
        XCTAssertTrue(restore.waitForExistence(timeout: BaseRobot.defaultTimeout),
                      "RESTORE should be present on '\(backupName)'")
        XCTAssertFalse(restore.isEnabled, "RESTORE should be gated (disabled)")
        return self
    }

    @discardableResult
    func assertRestoreConfirmVisible() -> ProjectsRobot {
        assertTagDisplayed(TestTags.PROJECTS_RESTORE_CONFIRM_DIALOG)
        return self
    }

    // ── Scroll helper ──

    private func scrollToSlot(_ nodeId: Int) {
        let list = element(TestTags.PROJECTS_SLOT_LIST)
        _ = list.waitForExistence(timeout: BaseRobot.defaultTimeout)
        let slot = element(TestTags.projectSlot(nodeId))
        var attempts = 0
        while attempts < 8 {
            if slot.exists && slot.isHittable { return }
            list.swipeUp()
            attempts += 1
        }
    }
}
